### DISCUSSION TITLE
Question: 3 minute rounding - duration does not match end

### DISCUSSION THREAD

**Discussion by limitedAtonement** [kimai/kimai#3100] [Questions & Answers]
**Describe the bug**
Rounding time entries to three minutes causes incorrect entries.

**To Reproduce**
1. Visit `/en/admin/system-config/`
   1. Set "Rounding of the start time in minutes" to "0"
   1. Set "Rounding of the end time in minutes" to "0"
   1. Set "Rounding of the duration in minutes" to "3"
1. Click Save
1. Before 21:17, make an entry that starts at 21:00 with no end date (so that it is in progress)
2. At 21:17, click "Stop" in the actions dropdown to the right.
2. Notice that although the duration was properly rounded to a multiple of three minutes, the stop time was not rounded. Calculating the duration in your head, this is a 17-minute entry, but the duration shows 18 minutes:

![image](https://user-images.githubusercontent.com/7562490/118063640-83d96d00-b367-11eb-8b5c-af474b49527d.png)

This is highly reproducible.

Also, this is clearly not just a presentation issue. I came across this issue because I wrote a script that reads from the database and puts the times into another timekeeping system. That other timekeeping system showed significantly less time than Kimai. Tracing down the problem, it was because I put the times into the other system using start and end times, not durations, and Kimai had not rounded the end times as expected.

**Additional context**
 - Kimai version: https://hub.docker.com/layers/kimai/kimai2/apache-debian-master-prod/images/sha256-8dd17cc2f556501c1e193172d3805b93f21e0a4304349621fd3fa0564c83cdb2
 - Browser: Chromium 90.0.4430.85 Arch Linux

**Comment by kevinpapst** [comment]
More details required. What is the actual result and what did you expect?

**Comment by limitedAtonement** [comment]
Understood. I don't know exactly what happened, only that after two days of recording times, the data were slightly messed up compared to another time tracking system. I'll get better reproduction instructions for you.

**Comment by pbksol** [comment]
Just my 2ct… Correct me if I'm wrong but that is the intended behaviour. If you set "Rounding of the end time in minutes" to "0" and "Rounding of the duration in minutes" to "3" only the duration is rounded. Rounding is off for the end time. If you want to round the end time you have to set a rounding parameter greater than 0.

**Comment by limitedAtonement** [comment]
@pbksol Just to verify explicitly, this kind of entry looks okay to you in this case?
![image](https://user-images.githubusercontent.com/7562490/118528195-0ec3ba00-b710-11eb-9789-a40140ed52dd.png)

Also note, that when I edit the timesheet and click in the various fields, the end time gets adjusted without me typing anything. Kimai seems to do its best to make sure "end time" - "start time" = "duration" in every context, which makes this particular case of rounding the durations seem out of place.

**Comment by pbksol** [comment]
Yep that's perfectly fine. The real duration was 17 minutes. But due to rounding the customer has to pay for 18 minutes.
This makes sense from a business perspective, too. This is the only way to create internal reports on how much the billed time differs from the actual work time.
So yes, that is exactly what I expect. If the customers should not see this information, only show start times and durations.
Or how we do it: use the start time, add the duration and all times match. We make this information available on a customer portal.

/edit: here's a possible solution. I did not test every case but the results seems to fulfil the requirements:
<img width="484" alt="Bildschirmfoto 2021-05-17 um 19 25 22" src="https://user-images.githubusercontent.com/8238713/118531130-c9ba7a80-b745-11eb-8048-4fbfd8011ba7.png">
Give it a try :-)

**Comment by limitedAtonement** [comment]
Your explanation is compelling even though it doesn't match my use case which is: I would like to bill in three-minute increments (to avoid rounding dollar amounts at billing time) and I am happy for "hours billed" to match exactly "dollars billed".

Also, if your explanation is the Kimai answer, there is likely a programming error in the timesheet. In this case, when the duration doesn't match "end time" - "start time", kimai updates the end time (undesirably in your case I believe!) to match the duration.

I will try your suggestion, thank you!

**Comment by limitedAtonement** [comment]
@pbksol Your suggestion (round starting AND ending times to multiples of three minutes) is working as expected, and isn't as annoying as I thought it would be.

Still, the ideal functionality in my mind is "Round the duration to a multiple of 3 minutes and adjust the end time to match". It appears to me that the "Rounding of the duration in minutes" setting was intended to be implemented this way. Your understanding is that the "Rounding the duration..." setting is STRICTLY for rounding the duration (WITHOUT adjusting the end time at all). Either way, there's still an error in Kimai:

1. If I'm right, the end time should be automatically rounded to match the properly rounded duration.
1. If you're right, the end time should not change when the timesheet is opened then saved.

The second error is captured in this screen recording: https://user-images.githubusercontent.com/7562490/120197275-cb2c7e00-c1ee-11eb-9db2-2af417ec99ea.mp4 in which I

1. Started and stopped a new time entry resulting in a "end time - start time" = 1 minute, with duration rounded to 3 minutes
1. Clicked the time entry to open the "edit time entry" modal
2. Clicked "save" at the bottom to close the "edit time entry" modal
1. Observe that the end time changed to match the rounded duration.

I would be happy if we could skip steps two and three, and if I understand correctly, you would be happy if step four did not occur.

Am I understanding correctly?

**Comment by kevinpapst** [comment]
The times in the UI change, because of the "convenience features" within the begin/end/duration fields.
This might be wrong in some edge cases - 80-20 rule applies here.

If someone wants to contribute to the javascript code and improve the behaviour, I'd be glad to have a look at it!
